sse/event_source.go

```diff
@@ -38,23 +38,47 @@ type EventSource struct {
 
 	currentReadCloser *ReadCloser
 	lastEventID       string
-	retryInterval     time.Duration
 	lock              sync.Mutex
 
-	closeOnce *sync.Once
-	closed    chan struct{}
+	closeOnce   *sync.Once
+	closed      chan struct{}
+	retryParams RetryParams
+}
+
+type RetryParams struct {
+	RetryInterval time.Duration
+	MaxRetries    uint16
 }
 
 func NewEventSource(client *http.Client, defaultRetryInterval time.Duration, requestCreator func() *http.Request) *EventSource {
+	retryParams := RetryParams{
+		RetryInterval: defaultRetryInterval,
+	}
+	return createEventSource(client, retryParams, requestCreator)
+}
+
+func createEventSource(client *http.Client, retryParams RetryParams, requestCreator func() *http.Request) *EventSource {
 	return &EventSource{
 		client:        client,
 		createRequest: requestCreator,
 
-		retryInterval: defaultRetryInterval,
+		closeOnce:   new(sync.Once),
+		closed:      make(chan struct{}),
+		retryParams: retryParams,
+	}
+}
+
+func ConnectWithParams(client *http.Client, retryParams RetryParams, requestCreator func() *http.Request) (*EventSource, error) {
```

Comment:
I am faced some issues going down the `net.Dialer` approach for this. Unlike `net.Dialer`, which has no pointer fields, in our case `http.Client` is a pointer (as `http.DefaultClient`) and hence just creating a var of that type (as done in `net.Dial` method)  is not sufficient, which means we need to have a `DefaultConfig` that uses `http.DefaultClient` as its `Client` field. Also, there is really no good default for `RequestCreator`. What do we default the request to?
Anyways, I am pushing changes that are close to what you think. Please take a look and we can discuss more.